Prepare CIL for concurrency

arkocal opened the pull request:
Replace references with DLS to enable multiple domains.

FelixKrayer committed:
make Cil.Pretty threadsafe usind Domain.DLS

arkocal committed:
Merge branch 'pretty-dls' of github.com:FelixKrayer/goblint-cil into develop

arkocal committed:
Add domain_shims for Ocaml4 compat

michael-schwarz commented:
Strange, it seems it builds but the CI tests still fail for OCaml <5.0? 

(For debugging, the note on how to run a single test under dune from the README may be helpful here.)

arkocal commented:
> Strange, it seems it builds but the CI tests still fail for OCaml <5.0?
> 
> (For debugging, the note on how to run a single test under dune from the README may be helpful here.)

I have tried, but could not find a solution yet.

Considering that older versions dont even build (due to dependencies) it might also make sense to drop domain_shims and keep this for an OCaml 5 Version in the future.

arkocal committed:
Fix test 318/404 longBlock

By adapting test script to use domain_shims (like in 13fc0203ab5a2f466929b2fe17a0d6d6b3b0f595)

sim642 commented on src/ocamlutil/pretty.mli:
```diff
@@ -294,25 +297,25 @@ val withPrintDepth : int -> (unit -> unit) -> unit
 
 (** Whether to rebalance doc before printing it to avoid stack-overflows *)
-val flattenBeforePrint : bool ref
+val flattenBeforePrint : bool refDLS
```
There's a mixture of `ref`s here:
1. Some are global state for some internal printing state, which should be DLS for printing to work right in parallel.
2. Others are more like global options which aren't set by the internal logic but there to be configured by the user (of the library). I'm not sure these should be DLS because it's something you would set at the beginning and likely not modify concurrently. It might also be a bit weird that changing the options in one domain (e.g. the main one), it's suddenly different in all other domains. So one would have to insert some extra logic to each domain spawn to set the options again but for that domain now.

arkocal replied:
@sim642 I believe these are the global options. Is there anything I oversee or should stay thread specific?
```
val printDepth   : int refDLS

val printIndent  : bool refDLS  (** If false then does not indent *)


(** If set to [true] then optional breaks are taken only when the document 
    has exceeded the given width. This means that the printout will looked 
    more ragged but it will be faster *)
val fastMode  : bool refDLS

val flushOften   : bool refDLS  (** If true the it flushes after every print *)

(** Whether to rebalance doc before printing it to avoid stack-overflows *)
val flattenBeforePrint : bool refDLS
```

sim642 commented on src/ocamlutil/pretty.ml:
```diff
@@ -241,33 +249,33 @@ type align =
 
 (* We use references to avoid the need to pass data around all the time *)
-let aligns: align list ref =  (* The current stack of active alignment marks,
+let aligns: align list refDLS =  (* The current stack of active alignment marks,
```
Here `aligns` is DLS, but also every field of `align` is made DLS. The latter seems excessive then, or would individual elements actually be exchanged between domains somehow?

arkocal replied:
Addressed in 84c012bd01dd27e3990acbb8cc0d0ec3c7b24f15 . When mixing `ref` and `refDLS` we cannot override `:=` anymore, but I think it does not hurt to make uses of DLS explicit.

sim642 commented on src/ocamlutil/pretty.ml:
```diff
@@ -456,16 +464,16 @@ let rec scan (abscol: int) (d: doc) : int =
+  let aligns: int list refDLS = refDLS [0] in (* A stack of alignment columns *)
 
-  let wantIndent = ref false in
+  let wantIndent = refDLS false in
```
Some things like this are local to a function. Do they actually escape and get transferred to other domains that they need DLS? If not, then a normal `ref` should be fine.

arkocal replied:
8837038

arkocal committed:
Use DLS only when necessary

arkocal committed:
Remove function local DLSRef in favor of refs

arkocal committed:
Make module wide settings refs instead of DLS again

arkocal committed:
Drop support for ocaml < 4.12

michael-schwarz commented on src/ciloptions.ml:
```diff
@@ -368,7 +368,7 @@ let options : (string * Arg.spec * string) list =
     Arg.Int (fun n -> Pretty.printDepth := n),
     ("<n> Set max print depth (default: " ^
-       string_of_int !Pretty.printDepth ^ ")");
+       string_of_int ! Pretty.printDepth ^ ")");
```
The space before deref is rather unusual in OCaml afaik.

michael-schwarz commented on src/ocamlutil/pretty.ml:
```diff
@@ -41,6 +41,16 @@
+module DLS =  Domain.DLS
+
+type 'a refDLS = 'a Domain.DLS.key
+(* let (!) = Domain.DLS.get *)
+(* let (:=) = Domain.DLS.set *)
```
This is outdated now, right?

michael-schwarz commented on src/ocamlutil/pretty.ml:
```diff
@@ -255,54 +265,56 @@ type align =
-  res.deltaFromPrev := abscol - !topAlignAbsCol;
-  topAlignAbsCol := abscol
+  DLS.set aligns (res :: (DLS.get aligns));
+  let (newdelta : int) = abscol - (DLS.get topAlignAbsCol) in
```
```suggestion
  let newdelta = abscol - (DLS.get topAlignAbsCol) in
```

Unless somehow needed for compilation.

michael-schwarz commented on src/ocamlutil/pretty.ml:
```diff
@@ -255,54 +265,56 @@ type align =
+  DLS.set aligns (res :: (DLS.get aligns));
+  let (newdelta : int) = abscol - (DLS.get topAlignAbsCol) in
+  res.deltaFromPrev := newdelta;
+  (* res.deltaFromPrev := (abscol - !topAlignAbsCol); *)
```
Why keep the old one around?

michael-schwarz commented on src/ocamlutil/pretty.mli:
```diff
@@ -296,23 +299,23 @@ val withPrintDepth : int -> (unit -> unit) -> unit
+val fastMode  : bool ref
 
-val flushOften   : bool ref  (** If true the it flushes after every print *)
+val flushOften   : bool ref(** If true the it flushes after every print *)
```
Having the doc comment here without space is a bit ugly.
```suggestion
val flushOften   : bool ref (** If true the it flushes after every print *)
```

arkocal committed:
Minor style fixes

sim642 commented on dune-project:
```diff
@@ -2,7 +2,7 @@
 (implicit_transitive_deps false)
 (generate_opam_files true)
-(version 2.0.5)
+(version 2.0.6)
```
The version number should remain the same for the time being. It's not a patch-level change either.

sim642 replied:
I think I want to make a last 2.0.6 release without this and have this in 2.1.0.

sim642 commented on src/dune:
```diff
@@ -68,7 +68,7 @@
  (modules main mainFeature ciloptions)
  (modes native)
- (libraries goblint-cil dynlink findlib)
+ (libraries goblint-cil dynlink findlib domain_shims)
```
The `main` executable doesn't use domain_shims directly though? I think only the goblint-cil library does, so the dependency comes transitively.

sim642 commented on src/ocamlutil/pretty.ml:
```diff
@@ -41,6 +41,14 @@
    structured text.
 *)
 
+(* redefine ref as Domain.DLS to avoid issues with Domains *)
```
This is no longer the case.

sim642 commented on src/ocamlutil/pretty.ml:
```diff
@@ -41,6 +41,14 @@
+
+type 'a refDLS = 'a Domain.DLS.key
+let refDLS vl = DLS.new_key (fun () -> vl)
+let incr rf = DLS.set rf (DLS.get rf + 1)
+let decr rf = DLS.set rf (DLS.get rf - 1)
```
Maybe also call these `incrDLS` and `decrDLS` for consistency?

sim642 commented on src/ocamlutil/pretty.ml:
```diff
@@ -456,32 +465,32 @@ let rec scan (abscol: int) (d: doc) : int =
                                            string *)
     (d: doc) =
-  let aligns: int list ref = ref [0] in (* A stack of alignment columns *)
+  let aligns: int list refDLS = refDLS [0] in (* A stack of alignment columns *)
```
I'm confused about this one. This is a local reference, but there's also a global one with the same name. How are these related?
By defining `aligns` locally, this function has no access to the global one, so it seems like this could be just a `ref` still. Or am I missing something?

sim642 commented on src/ocamlutil/pretty.ml:
```diff
@@ -255,54 +263,55 @@ type align =
-  aligns := res :: !aligns;
-  res.deltaFromPrev := abscol - !topAlignAbsCol;
-  topAlignAbsCol := abscol
+  DLS.set aligns (res :: (DLS.get aligns));
```
There's lots of unnecessary parentheses around `DLS.get` now all over this file.
